### DISCUSSION TITLE
How to make a move ordering system?

### DISCUSSION THREAD

**Discussion by manfromfms** [SebLague/Chess-Challenge#134] [Q&A]
Well i understand that it has to guess which move is better but how do i make it sort faster? I tried to make function for ordering but it was soooo slow

**Comment by stone50** [comment]
You could use the [OrderByDescending](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.orderbydescending?view=net-6.0) method. The way I'm doing it, I have a struct to contain a move and an associated evaluation.
```
struct MoveEval
{
   public Move Move;
   public float Eval;
}
```
Then I sort them like this:
```
// Assume that GetMoveEvals() returns an Enumerable of type MoveEval
MoveEval[] sortedMoveEvals = GetMoveEvals().OrderByDescending(moveEval => moveEval.Eval).ToArray();
```

**Reply by manfromfms** [reply]
Well it is slow. Maybe sth is wrong with my code but it seems to be working worse than without any sorts at all

**Comment by WhiteMouse1** [comment]
How do you measure the slowness?

**Reply by manfromfms** [reply]
Seconds per move

**Comment by mcthouacbb** [comment]
Take a look at the [CPW page for move ordering](https://www.chessprogramming.org/Move_Ordering) and [The rustic page for move ordering](https://rustic-chess.org/search/ordering/ordering.html)
Actually, go ahead and read everything on the second site since it's very helpful. It has a lot of techniques that are difficult to find elsewhere or aren't explicitly mentioned.

**Reply by mcthouacbb** [reply]
Also, if you try to order captures, make sure you use MVV_LVA(unless you have a different algorithm that takes into account the rest of the board, such as [Static Exchange Evaluation](https://www.chessprogramming.org/Static_Exchange_Evaluation). It might see counterintuitive, but Rook takes rook should be searched before pawn takes bishop. A lot of research has gone into move ordering in chess.
